### DISCUSSION TITLE
MIDI

### DISCUSSION THREAD

**Discussion by Moa13** [ctag-fh-kiel/ctag-tbd#52] [General]
Hi, a new TBD owner here and just being a fan of the device without any coding knowledge, the manual mentions briefly about possible MIDI parameters being exposed in the GUI, so is this more of a suggestive thing that could be created in a custom firmware?

**Comment by ctag-fh-kiel** [comment]
Hi there, Welcome to the TBD community :)

We worked on a MIDI implementation through the browser until about a year ago.
Everything is in the MIDI branch of the Git repository. The implementation allows one to use web midi (only Google Chrome supported) to learn control of web GUI parameters, which are then sent to TBD.

However, we stopped working on it as it requires https connection to TBD, which is a bit of a headache on the hardware side due to the ssl certificates.

Currently, there are no plans to make this available as release.
You're welcome to play around with it yourself (check out the MIDI branch, build and deploy), I am happy to support through the forum if needed...

**Reply by Moa13** [reply]
Thanks for the welcome! I’m also new to GitHub so if you could please point me out to the particular topic?
After making this topic I also saw Cuckoo’s thread where you are mentioning TBD MKII with usb MIDI, since the current TBD also has a USB connection will it also work on it? Nonetheless I’ll be the first in line for the MKII as well :)

**Reply by ctag-fh-kiel** [reply]
TBD mk2 is not yet finalized, it depends on how the community develops during 2022, more/other features might be requested and included. Last Superbooth, Instruments of Things showed a prototype, which includes the features mentioned in the other thread, i.e. USB MIDI, more CVs etc. There is no official release date.

On the other hand, the MIDI branch in the current repository uses your laptop with the TBD UI open in a Chrome browser and map MIDI control from any external MIDI controller connected to your laptop to the web UI in the Chrome browser. There is transmission latency involved, as all data goes through wifi: MIDI controller -> Laptop -> Chrome Browser -> TBD web UI -> WIFI -> TBD Hardware. Here the link to the branch https://github.com/ctag-fh-kiel/ctag-tbd/tree/midi

In order to build it, you could fork the whole repository and enable and adapt the Github continuous CI actions in your repo to your needs (e.g. starting a build upon push into MIDI branch). Then the build will be done in the cloud, no need to set up the whole tool chain locally...

Cheers :)

**Reply by Moa13** [reply]
Good to hear about the development of MKII! I’m basically starting from zero with how things work here @github, do I need any extra software for this build besides the browser? In my case I have been using Web MIDI through iPad on a browser app that allows Web MIDI in the same manner as Chrome on a laptop so I can communicate a MIDI controllers through it, with that also the iPad sitting next to TBD’s antenna which hopefully could limit the amount of latency,  would such setup be suitable to get the MIDI going?

**Reply by ctag-fh-kiel** [reply]
I haven't tried Web MIDI on an iPad, sound interesting...

**Reply by Moa13** [reply]
It works well for eurorack modules that have web based configuration UI like for instance the ES-9, https://apps.apple.com/nl/app/web-midi-browser/id953846217
So I was wondering if this could also be a way to set up MIDI with TBD? 
I tried to follow your instructions but didn’t come much further then forking the repository (now where the Fork is it says 21 and I can’t click on it anymore, so I guess that’s the start? 
Still not too sure what I should be doing next - I tried copying the code from the MIDI branch, then creating a new repository and importing the MIDI branch code, but that might be a wrong approach? I also couldn’t exactly find where I should enable and adapt the Github continuous CI. The stuff that I looked up about Continuous CI was a bit too all over the place for my understanding. My excuses for going on here with the most basic stuff but can you be kind to let me know if setting this up needs a higher skill with Github/coding or I still should be able to give it a go? Then I’ll know wether to pursue this any further, thanks ahead :)

**Reply by ctag-fh-kiel** [reply]
Sorry for the slightly late reply, in order to get the MIDI branch running with the latest set of plugins you would have to merge it with master. To be honest, I think that without some programming experience (C++ and JavaScript), this will be very hard to accomplish. Maybe someone from the community would be interest in helping. Currently, I can only occasionally assist in showing possible directions...

**Reply by Moa13** [reply]
thanks for the insight, In the meantime I'll patiently await untill if someone might see this post and otherwise I'm already getting a bunch out of TBD!
